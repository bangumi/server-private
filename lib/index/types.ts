export enum IndexType {
  User = 0,
  Public = 1,
  Award = 2,
}

export enum IndexRelatedCategory {
  Subject = 0,
  Character = 1,
  Person = 2,
  Ep = 3,
}
