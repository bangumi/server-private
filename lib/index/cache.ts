export function getSlimCacheKey(id: number): string {
  return `idx:v4:slim:${id}`;
}
