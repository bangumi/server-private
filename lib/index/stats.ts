import { DateTime } from 'luxon';

import { db, op, schema } from '@app/drizzle';
import { IndexRelatedCategory } from '@app/lib/index/types.ts';
import { SubjectType } from '@app/lib/subject/type.ts';
import type { IIndexStats } from '@app/lib/types/res';
import { decode } from '@app/lib/utils';

interface IInnerIndexStats {
  '1'?: number;
  '2'?: number;
  '3'?: number;
  '4'?: number;
  '6'?: number;
  character?: number;
  person?: number;
  ep?: number;
  blog?: number;
  group_topic?: number;
  subject_topic?: number;
}

export async function updateIndexStats(indexId: number) {
  const now = DateTime.now().toUnixInteger();
  await db.transaction(async (tx) => {
    const data = await tx
      .select({
        cat: schema.chiiIndexRelated.cat,
        type: schema.chiiIndexRelated.type,
        count: op.countDistinct(schema.chiiIndexRelated.sid),
      })
      .from(schema.chiiIndexRelated)
      .where(
        op.and(op.eq(schema.chiiIndexRelated.rid, indexId), op.eq(schema.chiiIndexRelated.ban, 0)),
      )
      .groupBy(schema.chiiIndexRelated.cat, schema.chiiIndexRelated.type);

    const stats: IInnerIndexStats = {};
    let total = 0;

    for (const { cat, type, count } of data) {
      total += count;
      if (cat === IndexRelatedCategory.Subject) {
        switch (type) {
          case SubjectType.Book: {
            stats['1'] = count;
            break;
          }
          case SubjectType.Anime: {
            stats['2'] = count;
            break;
          }
          case SubjectType.Music: {
            stats['3'] = count;
            break;
          }
          case SubjectType.Game: {
            stats['4'] = count;
            break;
          }
          case SubjectType.Real: {
            stats['6'] = count;
            break;
          }
        }
      } else {
        switch (cat) {
          case IndexRelatedCategory.Character: {
            stats.character = count;
            break;
          }
          case IndexRelatedCategory.Person: {
            stats.person = count;
            break;
          }
          case IndexRelatedCategory.Ep: {
            stats.ep = count;
            break;
          }
          case IndexRelatedCategory.Blog: {
            stats.blog = count;
            break;
          }
          case IndexRelatedCategory.GroupTopic: {
            stats.group_topic = count;
            break;
          }
          case IndexRelatedCategory.SubjectTopic: {
            stats.subject_topic = count;
            break;
          }
        }
      }
    }

    const statsString = JSON.stringify(stats);

    await tx
      .update(schema.chiiIndexes)
      .set({ stats: statsString, total, updatedAt: now })
      .where(op.eq(schema.chiiIndexes.id, indexId));
  });
}

export function parseIndexStats(stats: string): IIndexStats {
  const innerStats = decode(stats) as IInnerIndexStats;
  return {
    subject: {
      book: innerStats['1'],
      anime: innerStats['2'],
      music: innerStats['3'],
      game: innerStats['4'],
      real: innerStats['6'],
    },
    character: innerStats.character,
    person: innerStats.person,
    episode: innerStats.ep,
    blog: innerStats.blog,
    groupTopic: innerStats.group_topic,
    subjectTopic: innerStats.subject_topic,
  };
}
