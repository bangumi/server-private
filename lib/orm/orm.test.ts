import { expect, test } from 'vitest';

import { fetchPermission, fetchUser } from './index';

test('should fetch user', async () => {
  await expect(fetchUser(382951)).resolves.toMatchSnapshot();
});

test('should fetch permission', async () => {
  await expect(fetchPermission(0)).resolves.toMatchInlineSnapshot(`
    Object {
      "ban_post": true,
      "ban_visit": true,
    }
  `);
});

test('should fallback to empty permission', async () => {
  await expect(fetchPermission(10)).resolves.toMatchSnapshot();
});
