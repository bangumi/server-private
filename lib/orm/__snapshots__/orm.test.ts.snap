// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should fallback to empty permission 1`] = `
Object {
  "app_erase": true,
  "doujin_subject_lock": true,
  "ep_edit": true,
  "ep_move": true,
  "manage_user": true,
  "manage_user_group": true,
  "mono_erase": true,
  "mono_merge": true,
  "report": true,
  "subject_edit": true,
  "subject_lock": true,
  "subject_refresh": true,
  "subject_related": true,
  "user_list": true,
}
`;

exports[`should fetch user 1`] = `
Object {
  "groupID": 10,
  "id": 382951,
  "img": "000/38/29/382951.jpg?r=1571167246",
  "nickname": "树洞酱",
  "regTime": 1512262276,
  "sign": "treeholechan@gmail.com 密码:lovemeplease",
  "username": "382951",
}
`;
